README.md
# markdown-formatter

[中文文档](./README_CN.md)

## Introduction

This is a tool to improve the writing efficiency of markdown, not only provides a relatively uniform format for the `markdown` user, but also provides some snippets.

## Instructions

After the installation is complete, you may need to restart your vscode.

In any `markdown` standard file with a `.md` suffix, you can use `shift+option+f`(Macos) or `ctrl+shift+f`(window) to quickly format the code. 

![example.gif](https://raw.githubusercontent.com/sumnow/markdown-formatter/master/images/example.gif)

> PS: It does not fix your markdown syntax errors. for example, there is no space after `#` as a title, because in the code `#` is a usable character that is widely used as a comment or variable declaration.

## Feature

### Ignore file format

If you want to ignore the current file ( not to be formatted ), you can insert below the code before  or after the article

```code
<!-- /* md-file-format-disable */ -->
```

### Snippets part

**First of all, You must config like this in `settings.json`**

```js
"[markdown]": {
  // auto fill
  "editor.quickSuggestions": {
    "other": true,
    "comments": true,
    "strings": true
  },
}
```

#### import image

Typing `img` 

![img](./images/example-img.gif)

#### import table

Typing `tab` 

![img](./images/example-tab.gif)

#### import code

Typing `js` , `html` , `css` , `python` , `go` , `java` , or `code` 

![img](./images/example-code.gif)

#### import list

Typing `ul` or `ol` 

![img](./images/example-list.gif)

### Formatter part

#### symbols

* For `，,。;；！、？：` add a space after these symbols; 
* `，：；！“”‘’（）？。` , unify it as a half-width character(Optional); 
* `.!?` add a space after these symbols, if before uppercase letters or chinese; 
* Supports converting Chinese symbols into full-width symbols according to context, or converting English into half-width symbols; 
* a space before and after the back-quote, which wrapped by back-quote will not be formatted; 

#### line

* Empty a line for the title; 
* Empty a line for the code block; 
* The table is aligned automatically; 
* Empty a line for the link block; 
* erase duplicated line; 

#### table

![alt](./images/example-tab_format.gif)

#### list

##### unordered list

``` markdown

* TITLE
  + Title
    - title

```

![alt](./images/example-list_format.gif)


#### code

Format the code, by `js-beautify` tool, currently only the `javascript` , `html` and `css` language; 

##### code block

```lang
function sayHello() {console.log('hello')}
```

1. If `lang` is `js` or `javascript` or empty, it will be formatted according to js syntax.
2. If `lang` is `html` , it will be formatted according to html syntax.
3. If `lang` is `css` , it will be formatted according to css syntax.
4. If you don't use `js` usually, You can disable code block or code area autoformat code by setting the parameter `formatOpt` to false.

> The js formatting rules can be configured in `settings.json` .

> Code blocks can clearly mark the type of language, so it is recommended to use code blocks, you can configure `codeAreaToBlock` to set the conversion method.

##### code area

The default is not formatted. If `codeAreaToBlock` is any language name, such as `js` or `go`, it will be converted into a code block and then formatted.

![img](./images/example-code_area.gif)

**Tips: Empty lines in the code area cannot exceed one line.**


## Config

```js
{
  // 是否启用格式化
  "markdownFormatter.enable": true,
  // 将配置里的全角符号转化为半角符号, 例如 `，：；！“”‘’（）？。` 
  // 当设置为 '_' 或者 '' 的时候, 不转换符号
  // 当设置为 'auto' 的时候, 自动根据上下文转换符号
  "markdownFormatter.fullWidthTurnHalfWidth": "auto",
  // 转换代码块为代码区，默认为''（空字符）,不转换
  // 可以设置为任意合法英文
  // 只有设置成js或者javascript，才会按照js语言语法格式化
  "markdownFormatter.codeAreaToBlock": '',
  // 是否格式化代码
  // false: 不格式化代码
  "markdownFormatter.formatCodes" : true,
  // 配置js-beautify
  // {}: beautifyjs的配置设置
  // 配置 `js-beautify` , 可以参考[这里](https://github.com/beautify-web/js-beautify)
  "markdownFormatter.formatOpt": {},
  // 格式化无序列表的符号 
  "markdownFormatter.formatULSymbol": true,
  // * > + > -
  "markdownFormatter.formatULSymbolOpt": {'tag': ["*", "+", "-"]},
  // 各种符号后是否需要空格
  "markdownFormatter.spaceAfterFullWidthOrHalfWidth": 'half',
  // 在文件首部显示时间
  "markdownFormatter.displayTime" : false,
  // 是否格式化表格
  "markdownFormatter.formatTable" : false,
  // 代表中文是英文的二倍宽度,用于表格对齐
  "markdownFormatter.formatTableOpt" : {'chineseCharterWidth': 2},
}
```

How to config `beautifyjs` , you can click[here](https://github.com/beautify-web/js-beautify)

### configuration example

You can refer to my configuration:

```js
// settings.json
// markdown-formatter conf
// Convert the code area of unnamed language into code block according to js type
// "markdownFormatter.codeAreaToBlock": "js",
// or not
"markdownFormatter.codeAreaToBlock": "",
// format punctuation automatically
"markdownFormatter.fullWidthTurnHalfWidth": "auto",
// Chinese punctuation formatted to English
// "markdownFormatter.fullWidthTurnHalfWidth": "，：；！“”‘’（）？。",
"markdownFormatter.formatOpt": {
  "indent_size": 2
},

"markdownFormatter.formatULSymbol": true,
"markdownFormatter.formatULSymbolOpt": {'tag': ["*", "+", "-"]},
"markdownFormatter.formatTable" : true,
// if you use 2-width chinese font, otherwise not set
"markdownFormatter.formatTableOpt" : {'chineseCharterWidth': 2},

"[markdown]": {
  // auto save
  "editor.formatOnSave": false,
  // show space
  "editor.renderWhitespace": "all",
  // 
  "editor.quickSuggestions": {
    "other": true,
    "comments": true,
    "strings": true
  },
  // 
  "editor.snippetSuggestions": "top",
  "editor.tabCompletion": "on",
  // 
  "editor.acceptSuggestionOnEnter": "on",
  // 
  "editor.defaultFormatter": "mervin.markdown-formatter"
}
```


## Software version and development environment

#### develop version

VSCode version 1.29.1 (macOS Mojave)

#### test version

VSCode version 1.33.1 (macOS Mojave)

## Contact

If you have any ideas, please contact me.

If you know how to format other languages using front-end libraries, please let me know.

email: mydiamervin@gmail.com  or [here](https://github.com/sumnow/markdown-formatter/issues)


## Gratitude

Thank you for helping me improve this tool.

* [whidy](https://github.com/whidy)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/41)
* [jackfirth](https://github.com/jackfirth)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/35)
* [fruh](https://github.com/fruh)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/36)
* [ddejohn](https://github.com/ddejohn)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/35)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/31)
* [vio1etus](https://github.com/vio1etus)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/29)
* [zybieku](https://github.com/zybieku)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/27)
* [rowild](https://github.com/rowild)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/24)
* [tifDev](https://github.com/tifDev)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/28)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/23)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/22)
* [lartpang](https://github.com/lartpang)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/18)
* [lancerXXXX](https://github.com/lancerXXXX)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/20)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/16)
* [rgeorgiev583](https://github.com/rgeorgiev583)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/14)
* [Ardeshir81](https://github.com/Ardeshir81)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/13)
* [busyrat](https://github.com/busyrat) 
  + Feature(https://github.com/sumnow/markdown-formatter/issues/10)
* [lartpang](https://github.com/lartpang) 
  + Bug(https://github.com/sumnow/markdown-formatter/issues/15)
  + Feature(https://github.com/sumnow/markdown-formatter/issues/11)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/9)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/8)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/7)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/6)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/5)
* [iostalks](https://github.com/iostalks)
  + Feature(https://github.com/sumnow/markdown-formatter/issues/3)
* [zhuoyan](https://github.com/zhuoyan)
  + Bug(https://github.com/sumnow/markdown-formatter/issues/1)


> Sorting only according to the time of the question



CHANGELOG.md
<!--
Created: Mon Nov 21 2022 17:22:26 GMT+0800 (China Standard Time)
Modified: Fri Jun 09 2023 22:40:08 GMT+0800 (China Standard Time)
-->

# Log

[中文文档](./CHANGELOG_CN.md)

## 0.9.9

* add feature: table alignment

## 0.9.8

* Add unit tests
* Add tslint 

## 0.9.7

* Fix issue: code block appear error sometimes

## 0.9.6

* Update snippets

## 0.9.5

* Update `CHANGELOG.md`

## 0.9.4

* New Feature: Disable format file tag is optional
* Fix issue: [How to disable formatter with a file](hhttps://github.com/sumnow/markdown-formatter/issues/41)
* Update `CHANGELOG.md`

## 0.9.3

* Fix issue: [yaml code blocks get formatted ](https://github.com/sumnow/markdown-formatter/issues/49)
* Update `CHANGELOG.md`

## 0.9.2

* Fix issue: [Autoformat incorrectly adding spaces in code fence](https://github.com/sumnow/markdown-formatter/issues/35)
* Update `CHANGELOG.md`

## 0.9.1

* Fix issue: [格式化序号有bug #45](https://github.com/sumnow/markdown-formatter/issues/45)
* Fix issue: [Spaces being added forever between comma and snippet](https://github.com/sumnow/markdown-formatter/issues/48)
* Update `CHANGELOG.md`

## 0.9.0

* Update `CHANGELOG.md`

## 0.8.9

* Fix issue: [Formatting issue in shell/bash code block](https://github.com/sumnow/markdown-formatter/issues/38)
* Fix issue: [Space remove in case of list entry ending with code](https://github.com/sumnow/markdown-formatter/issues/36)
* Update `CHANGELOG.md`

## 0.8.8

* Fix issue: [错误地加转义符号](https://github.com/sumnow/markdown-formatter/issues/29)
* Update `CHANGELOG.md`

## 0.8.7

* Fix issue: [Formatter adds trailing space between backticks and newlines, which conflicts with Markdown linters](https://github.com/sumnow/markdown-formatter/issues/31)
* Update `CHANGELOG.md`

## 0.8.6

* Fix issue: There is no content between consecutive titles, and blank lines cannot be formatted correctly.
* Update `CHANGELOG.md`

## 0.8.5

* Fix bug: [[Error Format]: TypeError: Cannot read property 'forEach' of null](https://github.com/sumnow/markdown-formatter/issues/28)
* Update `CHANGELOG.md`

## 0.8.4

* Fix bug: [引用下的表格](https://github.com/sumnow/markdown-formatter/issues/20)
* Update `CHANGELOG.md`

## 0.8.3

* Remove notification
* add configuration `spaceAfterFullWidthOrHalfWidth`

* Update `CHANGELOG.md`

## 0.8.2

* Fix bug
* Update `CHANGELOG.md`

## 0.8.1

* Fix bug: [2 blank lines added at the end of file, conflicts with markdownlint](https://github.com/sumnow/markdown-formatter/issues/22)
* Fix bug: [能否不要删除code block](https://github.com/sumnow/markdown-formatter/issues/18)

## 0.8.0

* Fix bug: [It's not formatting properly lines that contain regexp](https://github.com/sumnow/markdown-formatter/issues/22)
* Fix bug: [Formatting adds a "*" to Markdown Rules](https://github.com/sumnow/markdown-formatter/issues/23)

## 0.7.9

* Fix bug: [列表接着一级标题格式化出错 #16](https://github.com/sumnow/markdown-formatter/issues/16)
* Update `CHANGELOG.md`

## 0.7.8

* Update `README.md`
* Update `CHANGELOG.md`

## 0.7.7

* Upload again

## 0.7.6

* Fix bug: [bad formatting for dots](https://github.com/sumnow/markdown-formatter/issues/13) 
* Update `CHANGELOG.md`

## 0.7.5

* Fix bug: [格式化错误](https://github.com/sumnow/markdown-formatter/issues/12) 
* Fix bug: The `>` formatted as link in code block cause error.
* Update `CHANGELOG.md`

## 0.7.4

* Fix bug: the code area will formatted failed sometimes.
* Update snippets discribution
* Update `CHANGELOG.md`

## 0.7.3

* Fix bug: the list formatted failed sometimes.
* Update `CHANGELOG.md`

## 0.7.2

* Fix bug: the code after list will formatted failed
* Update `CHANGELOG.md`

## 0.7.1

* Fix bug
* Update `CHANGELOG.md`

## 0.7.0

* Fix bug: the title will not after breakline, while `displayTime` is `true` .
* Update `CHANGELOG.md`

## 0.6.9

* Fix bug: the code wrapped by `*` , `~~` , `**` or `***` will not formatted as expected.
* Update `CHANGELOG.md`

## 0.6.8

* Update `README.md`

## 0.6.7

* Cut multiple spaces after the symbol to 1
* Add the feature of replacing the full corner with a comma
* Update `README.md`
* Update `CHANGELOG.md`

## 0.6.6

* New feature: display time before the article

## 0.6.5

* All blank lines before H1 had been removed
* Update `CHANGELOG.md`

## 0.6.4

* A blank line in the code block is treated as continuous
* Update `README.md`

* Update `CHANGELOG.md`

## 0.6.3

* Add new `Snippets` : 'time'  and optimizate others
* Update `CHANGELOG.md`

## 0.6.2

* Back quote after breakline will not after a space now.
* Update `CHANGELOG.md`

## 0.6.1

* Code area will not formatted if `codeAreaToBlock` as default 
* Update `README.md`

* Update `CHANGELOG.md`

## 0.6.0

* Remove property `codeAreaFormat`

* Update `README.md`

* Update `CHANGELOG.md`

## 0.5.9

* Fix bug: Code area will not formatted as expected.
* Optimization
* Update `CHANGELOG.md`

## 0.5.8

* Optimization
* Update `CHANGELOG.md`

## 0.5.7

* Fix bug: Convert code block to code area will failed sometimes
* Update `CHANGELOG.md`

## 0.5.6

* New feature: Serial number alignment in ordered list
* Update `CHANGELOG.md`

## 0.5.5

* Fix bug: [半角符号不能转换](https://github.com/sumnow/markdown-formatter/issues/5)
* Fix bug: [格式化操作会破坏 `markdown-all-in-one` 插件生成的目录](https://github.com/sumnow/markdown-formatter/issues/8)
* Fix bug: [格式化操作会破坏原有链接](https://github.com/sumnow/markdown-formatter/issues/9)
* Update `CHANGELOG.md`

## 0.5.4

* Fix bug: [关于嵌套的连接的格式化问题](https://github.com/sumnow/markdown-formatter/issues/7)
* Fix bug: [对于引用区域的格式化是否可以控制不要添加多余的行?](https://github.com/sumnow/markdown-formatter/issues/6)
* Update `CHANGELOG.md`

## 0.5.3

* Fix bug: Format HTML and CSS in code block
* Update `CHANGELOG.md`

## 0.5.2

* Fix bug: Code area will not formatted as expected.
* Update `CHANGELOG.md`

## 0.5.1

* New feature: Convert full-width symbols to half-width symbols by config `fullWidthTurnHalfWidth`

* Update `README.md`

* Update `CHANGELOG.md`

## 0.5.0

* Fix bug: code area next to each other will be formatted error

out/src/extension.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = exports.formatted = void 0;
const vscode = require("vscode");
const FormatList_1 = require("./components/FormatList");
const FormatTable_1 = require("./components/FormatTable");
const FormatPunctuation_1 = require("./components/FormatPunctuation");
const FormatCode_1 = require("./components/FormatCode");
const FormatLink_1 = require("./components/FormatLink");
// import { FormatHTML } from './components/FormatHTML';
const handlerTime_1 = require("./components/handlerTime");
let config = vscode.workspace.getConfiguration('markdownFormatter');
let fullWidthTurnHalfWidth = config.get('fullWidthTurnHalfWidth', 'auto');
let codeAreaToBlock = config.get('codeAreaToBlock', '');
let displayTime = config.get('displayTime', false);
let enable = config.get('enable', true);
var formatCodes = config.get('formatCodes', true);
let formatTable = config.get('formatTable', false);
let formatOpt = config.get('formatOpt', {});
let formatULSymbol = config.get('formatULSymbol', true);
let spaceAfterFullWidthOrHalfWidth = config.get('spaceAfterFullWidthOrHalfWidth', 'half');
vscode.workspace.onDidChangeConfiguration(_ => {
    config = vscode.workspace.getConfiguration('markdownFormatter');
    fullWidthTurnHalfWidth = config.get('fullWidthTurnHalfWidth', 'auto');
    codeAreaToBlock = config.get('codeAreaToBlock', '');
    displayTime = config.get('displayTime', false);
    enable = config.get('enable', true);
    formatTable = config.get('formatTable', false);
    formatCodes = config.get('formatCodes', true);
    formatOpt = config.get('formatOpt', {});
    formatULSymbol = config.get('formatULSymbol', true);
    spaceAfterFullWidthOrHalfWidth = config.get('spaceAfterFullWidthOrHalfWidth', 'half');
});
// console.log(vscode.window.activeTextEditor.options.tabSize)
// repalce symbol
const CHINESE_SYMBOL = `，、：；！“”‘’（）？。`;
const ENGLISH_SYMBOL = `,,:;!""''()?.`;
// chinese symbol
const CHINESE_CHARCTER_SYMBOL = `([\\u4e00-\\u9fa5])`;
const ENGLISH_CHARCTER_SYMBOL = `([A-Za-z])`;
const ELLIPSIS_EXP = /(\.\.\.)+/g;
// const SPACE_EXP = /\n+\ +\n+/g;
// breakline before the text 
const BEGIN_LINE_EXP = /^\n+/g;
// punctuation which need a space after it
// const PUNCTUATION_EXP = /([，,。；;！、？：])\ */g;
// const PUNCTUATION_ALL_EXP = /([，,。；;！、？：])\ */g;
const PUNCTUATION_CHINESE_EXP = /([，。；！、？：])\ */g;
const PUNCTUATION_ENGLISH_EXP = /([,;])\ */g;
// period which need a space after it
const PUNCTUATION_SPACIAL_ENGLISH_EXP = /([\.\!\?\:])([A-Z\u4e00-\u9fa5])/g;
// h1 symbol
const H1_EXP = /\n+(# [^\n]+)\n*/g;
// h2,h3,h4... symbol
const H_EXP = /\n+(##+ [^\n]+)\n/g;
// table
const TABLE_EXP = /((?:(?:[^\n]*?\|[^\n]*)\ *)?(?:\r?\n|^))(?:[^|]+)((?:\|\ *(?::?-+:?|::)\ *|\|?(?:\ *(?::?-+:?|::)\ *\|)+)(?:\ *(?::?-+:?|::)\ *)?\ *\r?\n)((?:(?:[^\n]*?\|[^\n]*)\ *(?:\r?\n|$))+)/g;
//back quote
const BACK_QUOTE_EXP = /\ *`([^`\n]+)`\ */g;
const BACK_QUOTE_AFTER_BREAKLINE_EXP = /\n\ `([^`\n]+)`\ /g;
// issue: https://github.com/sumnow/markdown-formatter/issues/48
const BACK_QUOTE_WITH_SPACE_EXP = /\ `([^`\n]+)`\ /g;
// image link
const IMG_EXP = /([^[])(\!\[[^\n]+\]\([^\n]+\))/g;
// split line 
const SPLIT_LINE_EXP = /\- \- \-( \-)*/g;
// list 
// const LIST_EXP = /(((?:\n)+(?: {4}|\t)*(?:\d+\.|\-|\*|\+) [^\n]+)+)/g;
const LIST_EXP = /((\n(?: {2}|\t)*(?:\d+\.|\-|\*|\+) [^\n]+)+)/g;
// const LIST_OL_EXP = /((\n(?: {2}|\t)*(\d+)\. [^\n]+)+)/g
const LIST_OL_LI_EXP = /(\n(?: {2}|\t)*)(\d+)(\. [^\n]+)/g;
// avoid - - - format to * - -
// const LIST_UL_ST_EXP = /\n(?:\-|\*|\+) ([^\n]+)/g;
const LIST_UL_ST_EXP = /\n(?:\-|\*|\+) ([^\n]+)/g;
const LIST_UL_ND_EXP = /\n(?: {2}|\t)(?:\-|\*|\+) ([^\n]+)/g;
const LIST_UL_TH_EXP = /\n(?: {2}|\t){2}(?:\-|\*|\+) ([^\n]+)/g;
// const NO_PERIOD_BACK_QUOTE_EXP = /\ *`([^.`\n]+)`\ */g;
// const NO_PERIOD_BACK_QUOTE_EXP1 = /\ *`([^`\n]*\.[^`\n]*)`\ */g;
// link 
const LINK_SPACE_EXP = /\n(>+) *([^\n]+)/g;
const LINK_EXP = /\n((>+[^\n]*\n)+)/g;
//href
const HREF_EXP = /(?:http|https|file|ftp):\/\/[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\*\+,;=.]+/g;
// duplicated line
const EXTRALINE_EXP = /\n\n+/g;
const END_LINE_EXP = /\n\n$/g;
// code block
// const CODE_AREA_EXP = /\n+((?:(?: {4}|\t)+[^\n\-\+\*][^\n]+\n*)+)/g;
// const CODE_AREA_EXP = /\n+((?:(?: {4}|\t)+(?!\d\.|\+|\-|\*)[^\n]+\n)+)/g;
// const CODE_AREA_EXP = /\n+((?:(?: {4}|\t)+([^\+\d\.\-\*])[^\n]+\n)+)/g
const CODE_AREA_EXP = /\n+((?:(?: {4}|\t)[^\n]*\n)+)/g;
// const CODE_AREAS_EXP = /(\n(?:(?: {4}|\t)[^\n]*\n))\n/g
// const CODE_AREA_EXP = /(?:(?: {4}|\t)+[^\n]+\n*)+/g;
// const CODE_EXP = /\n*```([\s\S]+?)```\n*/g;
const CODE_BLOCK_EXP = /\n*```(?: *)(\w*)\n([\s\S]+?)(```)+\n+/g;
// line-break
const LINE_BREAK_EXP = /\r\n/g;
const TIME_EXP = /(<!--\nCreated: [^\n]+\nModified: )[^\n]+(\n-->)(\n+)/g;
const DISABLE_EXP = /<!-- \/\* md-file-format-disable \*\/ -->/g;
const TAG_START_EXP = /<(?:[^\/])(?:[^"'>]|"[^"]*"|'[^']*')*[^\/]>/g;
const TAG_SINGLE_EXP = /<(?:[^\/])(?:[^"'>]|"[^"]*"|'[^']*')*\/>/g;
const TAG_END_EXP = /<\/(?:[^"'>]|"[^"]*"|'[^']*')*>/g;
const ITALIC_EXP = /\*\ (`[^`]+`)\ \*/g;
const BOLD_EXP = /\*\*\ (`[^`]+`)\ \*\*/g;
const ITALIC_BOLD_EXP = /\*\*\*\ (`[^`]+`)\ \*\*\*/g;
const LINE_THROUGH_EXP = /\~\~\ (`[^`]+`)\ \~\~/g;
// format sort: time -> punctution
function formatted(textP) {
    if (!enable) {
        return textP;
    }
    // let text = document.getText(range) + '\n\n'
    let text = '\n' + textP + '\n\n';
    const textLast = text;
    if (text.match(DISABLE_EXP)) {
        console.log('format current file disabled');
        return textP;
    }
    // format \r\n to \n,fix
    text = text.replace(LINE_BREAK_EXP, '\n');
    // handler time
    if (displayTime) {
        text = text.match(TIME_EXP) ? text.replace(TIME_EXP, `$1${handlerTime_1.handlerTime(new Date())}$2\n`) : `<!--\nCreated: ${handlerTime_1.handlerTime(new Date())}\nModified: ${handlerTime_1.handlerTime(new Date())}\n-->\n\n` + text;
    }
    try {
        // format PUNCTUATION_EXP
        text = new FormatPunctuation_1.FormatPunctuation(text).formatted({ fullWidthTurnHalfWidth, spaceAfterFullWidthOrHalfWidth, BACK_QUOTE_WITH_SPACE_EXP, CODE_BLOCK_EXP, CODE_AREA_EXP, HREF_EXP, LIST_OL_LI_EXP, PUNCTUATION_CHINESE_EXP, PUNCTUATION_ENGLISH_EXP, PUNCTUATION_SPACIAL_ENGLISH_EXP, CHINESE_SYMBOL, ENGLISH_SYMBOL, ENGLISH_CHARCTER_SYMBOL, CHINESE_CHARCTER_SYMBOL });
        text = new FormatLink_1.FormatLink(text).formatted({ LINK_SPACE_EXP, LINK_EXP, CODE_BLOCK_EXP, TABLE_EXP });
        if (formatTable) {
            // handler table
            text = new FormatTable_1.FormatTable(text).formatted({ TABLE_EXP, LINK_EXP, CODE_BLOCK_EXP, CODE_AREA_EXP });
        }
        // handler js
        text = new FormatCode_1.FormatCode(text).formatted({ formatCodes, formatOpt, codeAreaToBlock, CODE_BLOCK_EXP, LIST_EXP, CODE_AREA_EXP, H1_EXP, BACK_QUOTE_EXP });
        // handler list
        text = new FormatList_1.FormatList(text).formatted({ formatULSymbol, LIST_EXP, LIST_UL_ST_EXP, LIST_UL_ND_EXP, LIST_UL_TH_EXP, LIST_OL_LI_EXP, SPLIT_LINE_EXP, CODE_BLOCK_EXP, CODE_AREA_EXP });
        // text = new FormatHTML(text).formatted({TAG_START_EXP,TAG_SINGLE_EXP,TAG_END_EXP})
        // text = text.replace(BACK_QUOTE_EXP, ' `$1` ')
        // remove space in `something`+space+breakline
        // https://github.com/sumnow/markdown-formatter/issues/36
        text = text.replace(/` \n+/g, '`\n\n');
        text = text.replace(BACK_QUOTE_AFTER_BREAKLINE_EXP, '\n`$1` ');
        text = text.replace(H_EXP, '\n\n' + '$1' + '\n\n');
        // text = text.replace(H1_EXP, '$1' + '\n\n')
        text = text.replace(IMG_EXP, '$1\n\n' + '$2' + '\n\n');
        text = text.replace(CODE_BLOCK_EXP, '\n\n```' + '$1\n$2' + '```\n\n');
        // text = text.replace(LINK_EXP, '\n\n' + '$1' + '\n\n')
        // text = text.replace(LINK_SPACE_EXP, '\n' + '$1 $2')
        text = text.replace(EXTRALINE_EXP, '\n\n');
        text = text.replace(ITALIC_BOLD_EXP, `***$1***`);
        text = text.replace(BOLD_EXP, '**$1**');
        text = text.replace(ITALIC_EXP, '*$1*');
        text = text.replace(LINE_THROUGH_EXP, '~~$1~~');
        // clear breakline
        text = text.replace(BEGIN_LINE_EXP, '');
        // decrease end line
        // https://github.com/sumnow/markdown-formatter/issues/24
        text = text.replace(END_LINE_EXP, '\n');
    }
    catch (e) {
        text = textLast;
        vscode.window.showInformationMessage(`[Error Format]:${e} \n you can ask for help by https://github.com/sumnow/markdown-formatter/issues`);
    }
    return text;
}
exports.formatted = formatted;
// let textLast = ''
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    context.subscriptions.push(vscode.languages.registerDocumentFormattingEditProvider('markdown', {
        provideDocumentFormattingEdits(document, options, token) {
            if (!enable) {
                return void 0;
            }
            const result = [];
            const start = new vscode.Position(0, 0);
            const end = new vscode.Position(document.lineCount - 1, document.lineAt(document.lineCount - 1).text.length);
            const range = new vscode.Range(start, end);
            let text = formatted(document.getText(range));
            result.push(new vscode.TextEdit(range, text));
            // vscode.window.showInformationMessage('Formatted text succeeded!');
            return result;
        }
    }));
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "markdown-formatter" is now active!');
    // The command has been defined in the README.md file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in README.md
    // let disposable = vscode.commands.registerCommand('extension.sayHello', () => {
    // The code you place here will be executed every time your command is executed
    // Display a message box to the user
    // vscode.window.showInformationMessage('Hello World!');
    // });
    // context.subscriptions.push(disposable);
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map